### ISSUE TITLE
Failure on Mac El Capitan

### ISSUE THREAD

**Issue by BarrySmith** [LLNL/spack#237]
1) doesn't find clang compilers

~/Src/spack (develop=) arch-hwloc
$ which clang
/usr/bin/clang
~/Src/spack (develop=) arch-hwloc
$ ./bin/spack compilers
==> Available compilers
-- gcc ----------------------------------------------------------
gcc@5.2.0  gcc@4.2.1
~/Src/spack (develop=) arch-hwloc

2) Cannot install cmake

$ ./bin/spack install cmake
==> Installing cmake
==> Already downloaded /Users/barrysmith/Src/spack/var/spack/stage/cmake-3.4.0-3ji5lddqvejw2y75nf2ebrcqlf6w2htw/cmake-3.4.0.tar.gz.
==> Already staged cmake in /Users/barrysmith/Src/spack/var/spack/stage/cmake-3.4.0-3ji5lddqvejw2y75nf2ebrcqlf6w2htw.
==> No patches needed for cmake.
==> Building cmake.
==> Error: Command exited with status 6:
./configure --prefix=/Users/barrysmith/Src/spack/opt/spack/darwin-x86_64/gcc-5.2.0/cmake-3.4.0-3ji5lddqvejw2y75nf2ebrcqlf6w2htw --parallel=8

See build log for details:
  /private/var/folders/c1/ldz_dt8n2r3dtwv_chp5pfr40000gn/T/spack-stage/spack-stage-OFjo7A/cmake-3.4.0/spack-build.out

/Users/barrysmith/Src/spack/var/spack/packages/cmake/package.py:46, in install:
     44       def install(self, spec, prefix):
     45           configure('--prefix='   + prefix,
  >> 46                     '--parallel=' + str(make_jobs))
     47           make()
     48           make('install')
==> Error: Installation process had nonzero exit code.
~/Src/spack (develop=) arch-hwloc
$ cat /private/var/folders/c1/ldz_dt8n2r3dtwv_chp5pfr40000gn/T/spack-stage/spack-stage-OFjo7A/cmake-3.4.0/spack-build.out
==> ./configure --prefix=/Users/barrysmith/Src/spack/opt/spack/darwin-x86_64/gcc-5.2.0/cmake-3.4.0-3ji5lddqvejw2y75nf2ebrcqlf6w2htw --parallel=8
---------------------------------------------
CMake 3.4.0, Copyright 2000-2015 Kitware, Inc.
---------------------------------------------
Error when bootstrapping CMake:
Cannot find appropriate C compiler on this system.
Please specify one using environment variable CC.
See cmake_bootstrap.log for compilers attempted.

---------------------------------------------
Log of errors: /private/var/folders/c1/ldz_dt8n2r3dtwv_chp5pfr40000gn/T/spack-stage/spack-stage-OFjo7A/cmake-3.4.0/Bootstrap.cmk/cmake_bootstrap.log
---------------------------------------------
~/Src/spack (develop=) arch-hwloc
$

**Comment by tgamblin** [comment]
@BarrySmith: I can look at the compiler search functionality.  My suspicion is that the way `clang` formats its version info on Mac OS X.  Can you send me the output of `clang --version`?

You can see what Spack is looking for here:
https://github.com/LLNL/spack/blob/develop/lib/spack/spack/compilers/clang.py#L42

RE: the actual compile issues, that may be because there's no valid C compiler in `~/.spack/compilers.yaml`.  You mentioned that the compiler finder built you separate `gcc` entries for `gfortran` and `gcc` because they have separate versions.  We could potentially add some smarts to at least warn/ask users when that happens.  Can you paste your `compilers.yaml`, and see if this works if you merge your two separate `gcc` entries into one?

**Comment by BarrySmith** [comment]
> On Dec 15, 2015, at 5:18 PM, Todd Gamblin <notifications@github.com> wrote:
> 
> @BarrySmith: I can look at the compiler search functionality. My suspicion is that the way clang formats its version info on Mac OS X. Can you send me the output of clang --version?
> 
> You can see what Spack is looking for here:
> https://github.com/LLNL/spack/blob/develop/lib/spack/spack/compilers/clang.py#L42
> 
$ clang --version
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin15.2.0
Thread model: posix


  Clearly you guys need access to one or more Macs running different versions of Mac OS and Xcode. I'd volunteer the 2 retired Macs I've bought over the years for my son but surely LLNL with well over 1 billion dollars a year annual budge could afford them itself? Also we've found with PETSc configure that the more information you can log into a single file that then can be sent to the spack developers to debug what went wrong the better. 

  Sorry for being so cynical. We'd be happy to try out spack on all the systems we test PETSc on if you are interested? 

> RE: the actual compile issues, that may be because there's no valid C compiler in ~/.spack/compilers.yaml. You mentioned that the compiler finder built you separate gcc entries for gfortran and gcc because they have separate versions. We could potentially add some smarts to at least warn/ask users when that happens. Can you paste your compilers.yaml, and see if this works if you merge your two separate gcc entries into one?

  I'd suggest you add a spack function like 

spack add compiler family 

that prompts for c, C++ and Fortran compiler. Expecting users to edit .spack is a not a good public API.


   On a completely separate subject I am not sure putting .spack in the home directory while the rest of spack is handled off the directory where the spack files are is a good idea. Why not put .spack with the rest of spack and then I can have multiple complete versions if I want?


  Another completely separate subject; why so eager to download the crappy blas/lapack from netlib when Apple has a perfectly fine one come with Xcode? Better to use the installed one when possible IMHO. The download version takes forever to build, why would users want to wait around for that?


$ bin/spack install petsc
==> Installing petsc
==> Installing hypre
==> Installing netlib-lapack
==> Installing cmake
==> Trying to fetch from https://cmake.org/files/v3.4/cmake-3.4.0.tar.gz
######################################################################## 100.0%
==> Staging archive: /Users/barrysmith/Src/spack/var/spack/stage/cmake-3.4.0-ob4uiusyhuaoegd7wkf2f5dcratvtadt/cmake-3.4.0.tar.gz
==> Created stage in /Users/barrysmith/Src/spack/var/spack/stage/cmake-3.4.0-ob4uiusyhuaoegd7wkf2f5dcratvtadt.
==> No patches needed for cmake.
==> Building cmake.
==> Successfully installed cmake.
  Fetch: 4.52s.  Build: 3m 11.35s.  Total: 3m 15.87s.
[+] /Users/barrysmith/Src/spack/opt/spack/darwin-x86_64/gcc-4.2.1/cmake-3.4.0-ob4uiusyhuaoegd7wkf2f5dcratvtadt
==> Installing netlib-blas
==> Trying to fetch from http://www.netlib.org/lapack/lapack-3.5.0.tgz
######################################################################## 100.0%
==> Staging archive: /Users/barrysmith/Src/spack/var/spack/stage/netlib-blas-3.5.0-suj526boqg3oiywk46w7muvrqa6di5mf/lapack-3.5.0.tgz
==> Created stage in /Users/barrysmith/Src/spack/var/spack/stage/netlib-blas-3.5.0-suj526boqg3oiywk46w7muvrqa6di5mf.
==> Patched netlib-blas
==> Building netlib-blas.
==> Successfully installed netlib-blas.
  Fetch: 3.11s.  Build: 33.57s.  Total: 36.68s.
[+] /Users/barrysmith/Src/spack/opt/spack/darwin-x86_64/gcc-4.2.1/netlib-blas-3.5.0-suj526boqg3oiywk46w7muvrqa6di5mf
==> Trying to fetch from http://www.netlib.org/lapack/lapack-3.5.0.tgz
######################################################################## 100.0%
==> Staging archive: /Users/barrysmith/Src/spack/var/spack/stage/netlib-lapack-3.5.0-wca7y32x765bhlzgmsjwqdym4xg5ia7c/lapack-3.5.0.tgz
==> Created stage in /Users/barrysmith/Src/spack/var/spack/stage/netlib-lapack-3.5.0-wca7y32x765bhlzgmsjwqdym4xg5ia7c.
==> No patches needed for netlib-lapack.
==> Building netlib-lapack.


 Barry

**Comment by mathstuf** [comment]
I'll try and look at this when I try out some other OS X fixes (currently running 10.10, but I do have access to some machines which have already upgraded to 10.11).

**Comment by eschnett** [comment]
To give people access to the pre-installed version of BLAS on OS X, there should probably be a pseudo-package that doesn't install anything, but only confirms that this system library is indeed present, and then sets up the respective environment variables.

**Comment by eschnett** [comment]
Regarding Fortran compilers on OS X: There is no pre-installed Fortran compiler. The usual way to install one is to use Macports on Homebrew.

So one way would be to have a Spack package that is a Homebrew wrapper to install certain basic software. I guess that would work fine. On the other hand, this leaves a strange aftertaste, since Homebrew is essentially just a system-specific version of Spack. Thus maybe using Spack to install gcc would be better, and then having a simple way to point Spack to the compiler that it just installed.

**Comment by tgamblin** [comment]
@eschnett: Using Spack to install the fortran compilers would be better.  You can do this by installing `gcc` right now (though I have not tested it on El Capitan).  Then doing `spack compiler add <gcc prefix>`.  So the registration of the compiler is manual but you can do it.  I would like it if we had something more automatic.

**Comment by tgamblin** [comment]
@BarrySmith:  Currently OS X is not that well supported in Spack because we have focused our testing on Linux, but I would like to make it more robust.

How do you install your compilers on OS X?  You say that you used mixed C and Fortran compiler versions -- is that typical?  Do others do that?  Or is that just because it's a pain to get the compilers built in the first place, and you only felt like building additional Fortran ones?

What if Spack installed the Fortran compilers automatically if they were missing?  Would that be better (see above)?

I am hesitant to add a `spack compiler add family` that prompts the user, mainly because I hate prompts.  I polled people around here and they share this opinion.  However, `spack compiler add` will take a prefix, e.g. I can do this:
```
    spack compiler add ~/macports
```
And that will add all the compilers detected in the macports `bin` directory.  `spack compiler add` without args will re-search the `PATH`, as well.  Is that not sufficient?  I don't think it's unreasonable to ask people to edit config file if they want mixed compiler versions.

I guess I would like Spack to make a better effort to assemble what's available into a usable compiler suite.  I suppose I could amend the auto-detection to try to do mixed versions **if** it doesn't find a consistent suite of compilers.  I would most likely have it emit a warning if that happens, telling the user they're getting mixed compilers.

Thoughts?

**Comment by tgamblin** [comment]
@BarrySmith @eschnett: We'll be adding support for builtin packages with PR #120.  Once that is in we can easily add something that pulls in Apple's builtin BLAS.

**Comment by BarrySmith** [comment]
@tgamblin Typically I use the Apple Xcode clang compilers for C++ and then a precompiled gfortran from some site like http://hpc.sourceforge.net  Note that generally I've had no problems mixing the gfortran from here with the Apple clang compilers. Even with upgrades in the OS or clang compilers.


I don't like to install the GNU gcc and g++ compilers; I prefer to use the clang.

**Comment by tgamblin** [comment]
@BarrySmith: Ok.  We have this issue for the Nag Fortran compilers as well -- it just came up on the mailing list that @ajstewart at ANL uses them with `gcc`.  The clang case is much more common.

I'll think about how to better support mixing compilers -- right now the family granularity is too coarse in many cases. You can do it but specifying is cumbersome.  If Spack built you some decent Fortran compilers would you feel the need to go to http://hpc.sourceforge.net/?

**Comment by BarrySmith** [comment]
It would need to be Fortran compilers that could work with the native clang compilers.

**Comment by tgamblin** [comment]
@BarrySmith: that's doable.

**Comment by eschnett** [comment]
The only free Fortran compiler comes with gcc. I just submitted a pull request to make gcc build on OS X with Spack. I tested this with the native Apple compilers. Such a gfortran should work fine together with Apple's C/C++ compilers. In fact, gcc's C and C++ compilers should also be compatible with Apples C/C++ compilers.

**Comment by adamjstewart** [comment]
The problem with finding Clang compilers should now be resolved as of #394. If you are still having problems with installing CMake, you may want to mark this issue as closed and create a new issue specific to CMake.

**Comment by davydden** [comment]
> How do you install your compilers on OS X? You say that you used mixed C and Fortran compiler versions -- is that typical? Do others do that? Or is that just because it's a pain to get the compilers built in the first place, and you only felt like building additional Fortran ones?

I just came across Spack and trying it on El Capitan OS-X. Indeed, the commmon usage case is `clang` for C/C++ and `gfortran` for everything else. On OS-X `clang` is so much faster than GNU C/C++ and the error messages for heavy templated code from my experience are much better.

For me, finding compilers seem to work
```
$ ./spack compilers
==> Available compilers
-- gcc ----------------------------------------------------------
gcc@5.3.0  gcc@4.2.1

-- clang --------------------------------------------------------
clang@7.0.2-apple
```
but the trick is to mix `gfortran` and `clang`. This and support for OS-X blas/lapack would be super helpful.

A word of caution about blas/lapack used with GNU fortran, if one needs single precision Fortran, then you have to use a wrapper library `veclibfort` https://github.com/mcg1969/vecLibFort (e.g. in `ScaLAPACK`). Otherwise one can link with `-llapack -lblas` as those are symlinked to `-framework Accelerate`. So this is the same as using system provided Blas/Lapack on Ubuntu. I suppose if someone has a paid `ifort`, then this is another story.

**Comment by adamjstewart** [comment]
@BarrySmith Are you still having problems finding clang or building gfortran?

**Comment by BarrySmith** [comment]
I got a new laptop recently which is a Spack virgin:

   Things were looking good until I got to this

######################################################################## 100.0%
==> Staging archive: /Users/barrysmith/Src/spack/var/spack/stage/hwloc-1.11.3-yxw7dwjuiil4u3yx4q6jtbq5o2x2453v/hwloc-1.11.3.tar.gz
==> Created stage in /Users/barrysmith/Src/spack/var/spack/stage/hwloc-1.11.3-yxw7dwjuiil4u3yx4q6jtbq5o2x2453v
==> No patches needed for hwloc
==> Building hwloc
==> Successfully installed hwloc
  Fetch: 1.62s.  Build: 26.24s.  Total: 27.86s.
[+] /Users/barrysmith/Src/spack/opt/spack/darwin-elcapitan-x86_64/gcc-6.1.0/hwloc-1.11.3-yxw7dwjuiil4u3yx4q6jtbq5o2x2453v
==> Warning: OpenMPI : FORTRAN compiler not found
==> Warning: OpenMPI : FORTRAN bindings will be disabled
==> Trying to fetch from http://www.open-mpi.org/software/ompi/v1.10/downloads/openmpi-1.10.3.tar.bz2


I have gfortran installed by brew in /usr/local 

Tell me what Spack diagnostics to run and send to you.

   Barry

**Comment by davydden** [comment]
@BarrySmith you need to hack mixed compilers (fortran + clang) manually at the moment, see, for example, this short how-to https://github.com/dealii/dealii/wiki/deal.II-in-Spack#mixing-gcc-and-clang-on-osx

**Comment by BarrySmith** [comment]
Ok, I've done that. Now I have 

$ spack compilers
==> Available compilers
-- gcc ----------------------------------------------------------
gcc@6.1.0  gcc@4.2.1

-- clang --------------------------------------------------------
clang@7.3.0-apple


  But I still get 

$ ./spack install petsc
==> Installing petsc
==> Installing hypre
==> Installing openmpi
==> hwloc is already installed in /Users/barrysmith/Src/spack/opt/spack/darwin-elcapitan-x86_64/gcc-6.1.0/hwloc-1.11.3-yxw7dwjuiil4u3yx4q6jtbq5o2x2453v
==> Warning: OpenMPI : FORTRAN compiler not found
==> Warning: OpenMPI : FORTRAN bindings will be disabled

Is it somehow using the gcc suite? How do I force it to use clang?

**Comment by adamjstewart** [comment]
This may also be the problem I reported in #1071 and is fixed by #1085. Once that gets merged, Spack should pick up gfortran again.

@BarrySmith: Can you run `spack config get compilers` and show us the output?

**Comment by davydden** [comment]
@BarrySmith: if you manually added fortran compiler to `clang@7.3.0-apple` then it should be fine. Can you paste the content of `~/.spack/compilers.yaml`?

**Comment by BarrySmith** [comment]
$ spack config get compilers
compilers:
- compiler:
    modules: []
    operating_system: elcapitan
    paths:
      cc: /usr/bin/clang
      cxx: /usr/bin/clang++
      f77: /usr/local/bin/gfortran
      fc: /usr/local/bin/gfortran
    spec: clang@7.3.0-apple
- compiler:
    modules: []
    operating_system: elcapitan
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: null
      fc: null
    spec: gcc@4.2.1
- compiler:
    modules: []
    operating_system: elcapitan
    paths:
      cc: /usr/local/bin/gcc-6
      cxx: /usr/local/bin/g++-6
      f77: null
      fc: null
    spec: gcc@6.1.0

$ cat ~/.spack/compilers.yaml
compilers:
- compiler:
    modules: []
    operating_system: elcapitan
    paths:
      cc: /usr/bin/clang
      cxx: /usr/bin/clang++
      f77: /usr/local/bin/gfortran
      fc: /usr/local/bin/gfortran
    spec: clang@7.3.0-apple
- compiler:
    modules: []
    operating_system: elcapitan
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: null
      fc: null
    spec: gcc@4.2.1
- compiler:
    modules: []
    operating_system: elcapitan
    paths:
      cc: /usr/local/bin/gcc-6
      cxx: /usr/local/bin/g++-6
      f77: null
      fc: null
    spec: gcc@6.1.0

**Comment by adamjstewart** [comment]
Yep, called it. The reason that gcc@6.1.0 didn't pick up /usr/local/bin/gfortran is due to #1071. For now you'll have to manually edit it, but once #1085 is merged you'll be good to go.

Spack defaults to the gcc compiler unless you tell it otherwise. This is because Spack is a Supercomputer PACKage manager, and supercomputers run Linux, which comes with gcc, not clang. If you want to try building with clang, you'll have to specify that during installation:
```
spack install hwloc %clang
```

**Comment by BarrySmith** [comment]
Just for fun I tried MPICH, it seems to have a different problem with Fortran.


$ ./spack install mpich
==> Installing mpich
==> Trying to fetch from http://www.mpich.org/static/downloads/3.2/mpich-3.2.tar.gz
######################################################################## 100.0%
==> Staging archive: /Users/barrysmith/Src/spack/var/spack/stage/mpich-3.2-uivdbjn3ozvvwrsym7y5nwbasyxc6mwu/mpich-3.2.tar.gz
==> Created stage in /Users/barrysmith/Src/spack/var/spack/stage/mpich-3.2-uivdbjn3ozvvwrsym7y5nwbasyxc6mwu
==> No patches needed for mpich
==> Building mpich
Traceback (most recent call last):
  File "/Users/barrysmith/Src/spack/lib/spack/spack/build_environment.py", line 494, in fork
    function()
  File "/Users/barrysmith/Src/spack/lib/spack/spack/package.py", line 958, in build_process
    self.install(self.spec, self.prefix)
  File "/Users/barrysmith/Src/spack/var/spack/repos/builtin/packages/mpich/package.py", line 93, in install
    self.filter_compilers()
  File "/Users/barrysmith/Src/spack/var/spack/repos/builtin/packages/mpich/package.py", line 114, in filter_compilers
    filter_file(env['F77'], self.compiler.f77, mpif77, **kwargs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/UserDict.py", line 23, in __getitem__
    raise KeyError(key)
KeyError: 'F77'
==> Error: Installation process had nonzero exit code : 256

**Comment by adamjstewart** [comment]
In both cases, the compiler you are using (gcc@6.1.0) does not have the Fortran compiler set correctly.

**Comment by BarrySmith** [comment]
Ok, running 

./spack install petsc %clang

now.  Working on OpenMPI and no complaint about missing Fortran.

 BTW: I found the -h message that said %compiler very unhelpful. It was not clear the %clang should come after the package. Perhaps the -h could give a couple examples showing exact usage.

  Also is there a way to set clang as the default compiler so I don't need to put it in for each install I do?

   Thanks

**Comment by davydden** [comment]
@BarrySmith i would say it is essentially the same problem -- no fortran compilers. However in MPICH python code somethere it is assumed that fortran compilers are not `null`.

Regarding the original problem, your `~/.spack/compilers.yaml` look good to me and it should work. Don't know which one is default, but try `spack install petsc%clang`. That way it should go with compiler within `clang@7.3.0-apple`.

**Comment by davydden** [comment]
> Also is there a way to set clang as the default compiler so I don't need to put it in for each install I do?

in `~/.spack/packages.yaml`
```
packages:
  all:
    compiler: [clang, gcc]
    providers:
      mpi: [openmpi]
      blas: [openblas]
      lapack: [openblas]
```

**Comment by adamjstewart** [comment]
>  BTW: I found the -h message that said %compiler very unhelpful. It was not clear the %clang should come after the package. Perhaps the -h could give a couple examples showing exact usage.

Yep, `spack help` is very useful. It does in fact state that the compiler should come after the package:
```
PACKAGE [CONSTRAINTS]

    CONSTRAINTS:
      @version
      %compiler  @compiler_version
      +variant
      -variant or ~variant
      =architecture
      [^DEPENDENCY [CONSTRAINTS] ...]
```
If you still aren't familiar with all of the ways you can run Spack, the documentation is really useful:
http://software.llnl.gov/spack/index.html

**Comment by BarrySmith** [comment]
Sorry, I am dumb, I found the below very confusing.


$ ./spack -h
usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-V] SUBCOMMAND ...

Spack: the Supercomputing PACKage Manager.

spec expressions:
  PACKAGE [CONSTRAINTS]

Why does it have the name of the package in the middle of the usage: information? I have no idea that ... was suppose to mean "spec expressions" I would have it print as 

usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-V] SUBCOMMAND PACKAGE [CONSTRAINTS]

and get rid of the confusing  lines 

spec expressions:
  PACKAGE [CONSTRAINTS]

Plus I would list the CONSTRAINTS options AFTER the subcommand choices (not before as you do) because they come after the SUBCOMMAND in the command line.

Please take this only as a suggest on how to make the documentation clearer, I am not criticizing your work or complaining to be a pain, I am try to indicate what I found confusing and ways that as a newbie I think it would be clearer to users (especially new ones). If spack succeeds it will make my life much simpler so I want you guys to succeed but to succeed you need to deal with picky users :-).  Also if spack is to succeed endusers cannot have to read any spack documentation.

**Comment by adamjstewart** [comment]
I think part of the problem is that Spack can do so many things. The following wouldn't make sense:
```
usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-V] SUBCOMMAND PACKAGE [CONSTRAINTS]
```
because many Spack commands don't take a package as an argument, only the ones that take a spec do. Perhaps it would be clearer if `spack help` mentioned that you can run it and specify a subcommand to get further information on how to use that subcommand. Although `spack help install` doesn't list the various constraints that could be used.

> Plus I would list the CONSTRAINTS options AFTER the subcommand choices (not before as you do) because they come after the SUBCOMMAND in the command line.

That's not a bad idea. Although it is nice to have the constraints near the top for quick reference.

> Also if spack is to succeed end users cannot have to read any spack documentation.

I'm not sure if this will ever be the case. We could spend most of our time generating a monolithic `spack help` that covers all possible use cases, but it wouldn't be as organized and as easy to understand as the online documentation.

@tgamblin Thoughts on reorganizing the `spack help` output?

**Comment by BarrySmith** [comment]
> because many Spack commands don't take a package as an argument, only the ones that take a spec do

   Hmm, I have no idea what a spec is and hope I never need to learn. If the package is optional then you would have 

> usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-V] SUBCOMMAND [PACKAGE [CONSTRAINTS]]

where the last ] goes before or after CONSTRAINTS depending on if constraints is only possible with a PACKAGE otherwise it is


> usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-V] SUBCOMMAND [PACKAGE] [CONSTRAINTS]

  Or do you also have 

> usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-V] SUBCOMMAND [options to a subcommand that does not take a package]?

  Then you might want to list two usages on two distinct lines.

> usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-V] SUBCOMMAND [PACKAGE [CONSTRAINTS]]

> usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-V] SUBCOMMAND [subcommand options]

  The more thought you guys put into organizing the help the less thought users require which is always a good thing.


  Barry

**Comment by adamjstewart** [comment]
> Hmm, I have no idea what a spec is and hope I never need to learn.

All of the following are specs:
```
mpich
mpich@3.2
mpich %clang
mpich +verbs %intel
```
And yes, we do have some subcommands which take other subsubcommands. But I like your suggestion of:
> usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-V] SUBCOMMAND [PACKAGE [CONSTRAINTS]]
 usage: spack [-h] [-d] [-D] [-k] [-m] [-p] [-v] [-V] SUBCOMMAND [subcommand options]

**Comment by tgamblin** [comment]
So, the output there is what's generated by `argparse`, and it's an exhaustive list of options.  Argparse automatically puts the less-than-wonderful `...` for variable arguments, which doesn't really tell you that Spack is expecting a package spec.

I have long wanted to replace that with something more minimal, where we show a simplified set of options and the most common subcommands, like `git` does.  We could then make the exhaustive list available via `spack help -a`, also like git does.

I like the suggestions above.  A `spec` is really just `[PACKAGE [CONSTRAINTS]]`.

**Comment by adamjstewart** [comment]
Anyway, it sounds like this issue is resolved. We should close this and make a new issue about how to make `spack help` more helpful.
